File: src/2025-02-23-deepclaude-try/java/exercises/practice/two-bucket/src/main/java/TwoBucket.java
import java.util.Queue;
import java.util.ArrayDeque;
import java.util.HashSet;
import java.util.Set;

class TwoBucket {
    private int totalMoves;
    private String finalBucket;
    private int otherLiters;

    private static class State {
        int a;
        int b;
        int moves;

        State(int a, int b, int moves) {
            this.a = a;
            this.b = b;
            this.moves = moves;
        }
    }

    TwoBucket(int bucketOneCap, int bucketTwoCap, int desiredLiters, String startBucket) {
        if (desiredLiters == 0) {
            throw new IllegalArgumentException("Desired liters cannot be zero");
        }

        Queue<State> queue = new ArrayDeque<>();
        Set<String> visited = new HashSet<>();

        int initialA = startBucket.equals("one") ? bucketOneCap : 0;
        int initialB = startBucket.equals("two") ? bucketTwoCap : 0;
        queue.add(new State(initialA, initialB, 1));
        visited.add(initialA + "," + initialB);

        boolean found = false;
        while (!queue.isEmpty()) {
            State current = queue.poll();

            if (current.a == desiredLiters || current.b == desiredLiters) {
                if (current.a == desiredLiters) {
                    finalBucket = "one";
                    otherLiters = current.b;
                } else {
                    finalBucket = "two";
                    otherLiters = current.a;
                }
                totalMoves = current.moves;
                found = true;
                break;
            }

            // Generate all possible moves
            generateStates(current, bucketOneCap, bucketTwoCap, startBucket, queue, visited);
        }

        if (!found) {
            throw new IllegalArgumentException("No solution possible");
        }
    }

    private void generateStates(State current, int cap1, int cap2, String startBucket, Queue<State> queue, Set<String> visited) {
        // Fill bucket1
        if (current.a < cap1) {
            processState(new State(cap1, current.b, current.moves + 1), startBucket, cap1, cap2, queue, visited);
        }

        // Fill bucket2
        if (current.b < cap2) {
            processState(new State(current.a, cap2, current.moves + 1), startBucket, cap1, cap2, queue, visited);
        }

        // Empty bucket1
        if (current.a > 0) {
            processState(new State(0, current.b, current.moves + 1), startBucket, cap1, cap2, queue, visited);
        }

        // Empty bucket2
        if (current.b > 0) {
            processState(new State(current.a, 0, current.moves + 1), startBucket, cap1, cap2, queue, visited);
        }

        // Pour bucket1 to bucket2
        int transfer12 = Math.min(current.a, cap2 - current.b);
        if (transfer12 > 0) {
            processState(new State(current.a - transfer12, current.b + transfer12, current.moves + 1), startBucket, cap1, cap2, queue, visited);
        }

        // Pour bucket2 to bucket1
        int transfer21 = Math.min(current.b, cap1 - current.a);
        if (transfer21 > 0) {
            processState(new State(current.a + transfer21, current.b - transfer21, current.moves + 1), startBucket, cap1, cap2, queue, visited);
        }
    }

    private void processState(State newState, String startBucket, int cap1, int cap2, Queue<State> queue, Set<String> visited) {
        // Check invalid state rule
        if ((startBucket.equals("one") && newState.a == 0 && newState.b == cap2) ||
            (startBucket.equals("two") && newState.b == 0 && newState.a == cap1)) {
            return;
        }

        String key = newState.a + "," + newState.b;
        if (!visited.contains(key)) {
            visited.add(key);
            queue.add(newState);
        }
    }

    int getTotalMoves() {
        return totalMoves;
    }

    String getFinalBucket() {
        return finalBucket;
    }

    int getOtherBucket() {
        return otherLiters;
    }
}

File: src/2025-02-23-deepclaude-try/python/exercises/practice/two-bucket/two_bucket.py
def measure(bucket_one, bucket_two, goal, start_bucket):
    # Input validation
    if bucket_one <= 0 or bucket_two <= 0:
        raise ValueError("Buckets must have positive capacities")
    if goal <= 0:
        raise ValueError("Goal must be a positive number")
    if start_bucket not in ('one', 'two'):
        raise ValueError("Start bucket must be 'one' or 'two'")
    if goal > max(bucket_one, bucket_two):
        raise ValueError("Goal cannot be larger than the largest bucket")

    # Check if goal is impossible based on GCD
    def gcd(a, b):
        while b:
            a, b = b, a % b
        return a

    if goal % gcd(bucket_one, bucket_two) != 0:
        raise ValueError("No solution possible")

    # Initialize starting state based on which bucket to fill first
    initial_state = (bucket_one, 0) if start_bucket == 'one' else (0, bucket_two)
    
    # Track visited states and moves
    visited = {initial_state}
    queue = [(initial_state, 1)]  # (state, moves)
    
    while queue:
        current_state, moves = queue.pop(0)
        current_one, current_two = current_state
        
        # Check if current state has the goal
        if current_one == goal:
            return (moves, 'one', current_two)
        if current_two == goal:
            return (moves, 'two', current_one)
            
        # Generate next possible states
        next_states = []
        
        # Fill operations
        if current_one < bucket_one:
            next_states.append((bucket_one, current_two))
        if current_two < bucket_two:
            next_states.append((current_one, bucket_two))
            
        # Empty operations
        if current_one > 0:
            next_states.append((0, current_two))
        if current_two > 0:
            next_states.append((current_one, 0))
            
        # Pour operations
        if current_one > 0 and current_two < bucket_two:
            pour = min(current_one, bucket_two - current_two)
            next_states.append((current_one - pour, current_two + pour))
        if current_two > 0 and current_one < bucket_one:
            pour = min(current_two, bucket_one - current_one)
            next_states.append((current_one + pour, current_two - pour))
            
        # Process next states
        for next_state in next_states:
            # Skip invalid states where starting bucket is empty and other is full
            if (start_bucket == 'one' and next_state[0] == 0 and next_state[1] == bucket_two) or \
               (start_bucket == 'two' and next_state[1] == 0 and next_state[0] == bucket_one):
                continue
                
            if next_state not in visited:
                visited.add(next_state)
                queue.append((next_state, moves + 1))
    
    raise ValueError("No solution possible")
